### PR TITLE
only run the background scanner for users that have unscanned files

### PR DESCRIPTION
**Pull request by icewind1991**
instead of looping trough all users and seeing if they have unscanned files, we do a single query to find all storages that need scanning and run trough the users for them

Signed-off-by: Robin Appelman <robin@icewind.nl>

### CODE REVIEW

**Review by rullzer — APPROVED** [review]

File: apps/files/lib/BackgroundJob/ScanFiles.php [hunk 1]
```diff
@@ -56,26 +59,30 @@ class ScanFiles extends \OC\BackgroundJob\TimedJob {
...
-	public function __construct(IConfig $config = null,
-								IUserManager $userManager = null,
-								IEventDispatcher $dispatcher = null,
-								ILogger $logger = null) {
+	public function __construct(
+		IConfig $config = null,
+		IUserManager $userManager = null,
+		IEventDispatcher $dispatcher = null,
+		ILogger $logger = null,
+		IDBConnection  $connection
```
Comment (nickvergessen) [inline apps/files/lib/BackgroundJob/ScanFiles.php:67]: can not be non-optional after optional parameters

File: apps/files/lib/BackgroundJob/ScanFiles.php [hunk 2]
```diff
@@ -87,6 +94,22 @@ protected function runScanner(IUser $user) {
...
 	}
 
+	/**
+	 * Find all storages which have unindexed files and return a user for each
+	 *
+	 * @return string[]
+	 */
+	private function getUsersToScan(): array {
+		$query = $this->connection->getQueryBuilder();
+		$query->select($query->func()->max('user_id'))
```
Comment (nickvergessen) [inline apps/files/lib/BackgroundJob/ScanFiles.php:104]: Max on string? You just want to get a function because of the group-by right?

**Comment by rullzer** [comment]
CI says no

**Comment by icewind1991** [comment]
redone the tests

File: apps/files/lib/BackgroundJob/ScanFiles.php [hunk 3]
```diff
@@ -87,6 +94,22 @@ protected function runScanner(IUser $user) {
...
+			->from('filecache', 'f')
+			->innerJoin('f', 'mounts', 'm', $query->expr()->eq('storage_id', 'storage'))
+			->where($query->expr()->lt('size', $query->createNamedParameter(0, IQueryBuilder::PARAM_INT)))
+			->groupBy('storage_id');
```
Comment (nickvergessen) [inline apps/files/lib/BackgroundJob/ScanFiles.php:108]:
```suggestion
			->groupBy('storage_id')
			->setMaxResults(self::USERS_PER_SESSION);
```

better save than sorry?

Comment (rullzer) [reply apps/files/lib/BackgroundJob/ScanFiles.php:108]: yep sounds sane

**Comment by rullzer** [comment]
ping @nickvergessen

**Review by nickvergessen — APPROVED** [review]
:elephant:

**Review by juliusknorr — APPROVED** [review]

**Pull request merged**